### ISSUE TITLE
Is there a relationship between dataset structure and ExampleProvider

### ISSUE THREAD

**Issue by drorhunvural** [gnina/libmolgrid#106]
The question may seem a bit long, but I'm trying to understand a very simple logic. That's why the question is actually **very very** simple for you.

I have a **train.types** file like below. 

```
2   12.531857  1.317891e+01  3.390248e+01 5bqh_protein_nowat.gninatypes
2   -5.128723  3.048503e+00  1.581255e+01 3c7p_protein_nowat.gninatypes
2   26.166937  2.637838e+01  5.408199e+01 3qw6_protein_nowat.gninatypes
0   11.532402 -1.458125e+01 -2.721688e+01 5m9d_protein_nowat.gninatypes
0    6.999138  3.286934e+00 -1.191319e+01 6mlp_protein_nowat.gninatypes
0    6.043935 -9.998359e-01  4.710859e+00 3kgq_protein_nowat.gninatypes
4  -17.732060  4.774173e+00  8.409735e+01 6g07_protein_nowat.gninatypes
4   45.587652  2.648775e+01  4.172926e+01 3pd8_protein_nowat.gninatypes
4  -18.591134 -1.809282e+01 -3.601412e+01 3zev_protein_nowat.gninatypes
1   -1.305433 -1.666860e+01 -7.233355e+00 5yz7_protein_nowat.gninatypes
1   10.138527  2.287561e+01  8.623280e+01 5gtr_protein_nowat.gninatypes
1   28.564483  1.619746e+02  1.473751e+02 6iiv_protein_nowat.gninatypes
3   21.603875  1.774174e+01  3.387129e+01 1sqo_protein_nowat.gninatypes
3  -11.421211  3.230708e+00  5.746333e+00 1gbq_protein_nowat.gninatypes
3   28.630292  8.507708e+00  2.745037e+01 2vio_protein_nowat.gninatypes
```

There are 5 different labels. These are 0,1,2,3,4. This is a classic classification problem. I have a folder containing all of these gninatypes called **allgninatypes**.  All xxx_protein_nowat.gninatypes files are in this folder. I also have a **pdb.molcache** file. This molcache file was created with the **create_caches2.py** script you shared.

I have a code like below. The code below includes both the model and all the codes in the train process. My model can't learn. The model keeps predicting the same label. I suspect something about what is the problem.

When I look at your example, you define Example provider as follows
`e = molgrid.ExampleProvider(data_root=datadir+"/structs",balanced=True,shuffle=True)`

Under the your struct folder, all proteins are gathered under separate folders. My file structure is not like this but my **train.types** (look at the top) file does not contain **xxx/xxx.gninatypes**. I have just **xxx.gninatypes.** 

I'm wondering if my software is accessing my gninatypes files and correctly identifying the example provider. Maybe you can ask me how did you create **gninatypes** files. I used a function similar to the one here (https://github.com/devalab/DeepPocket/blob/main/types_and_gninatyper.py)

My main question is to figure out why models don't learn. I tried on other complex models but couldn't solve it. Could I be wrong because of the things I mentioned above?

Finally, I give both the allgninatype folder and the molcache file to exampleprovider. Am I doing it right?

```
traintypes = '/train.types'
allgninatypes = '/allgninatypes'
recmolcache = '/pdb.molcache'



batch_size = 50

e = molgrid.ExampleProvider(data_root=allgninatypes,shuffle=True, recmolcache = train_recmolcache,stratify_receptor=True)
e.populate(traintypes)

gmaker = molgrid.GridMaker(binary=False)
dims = gmaker.grid_dimensions(e.num_types())
tensor_shape = (batch_size,)+dims

input_tensor = torch.zeros(tensor_shape, dtype=torch.float32, device='cuda',requires_grad=True)
float_labels = torch.zeros((batch_size,4), dtype=torch.float32, device='cuda')


```

**Comment by dkoes** [comment]
Balancing is for binary labels, what you want to do is stratify by the label value (set stratify_min to 0, stratify_max to 5, stratify_step to 1, and stratify_pos to 0). Make sure balanced is false.

**Comment by drorhunvural** [comment]
First of all, thank you very much for the reply. I understood the important points you emphasized. I followed what you recommend and my final ExampleProver was like below. 

`e = molgrid.ExampleProvider(data_root=allgninatypes, stratify_min = 0, stratify_max = 5, stratify_step=1, stratify_pos=0, shuffle=True, recmolcache = train_recmolcache,stratify_receptor=True, balanced = False)`

I think I'm making a logical error somewhere because the model's prediction and what it should actually look like are as follows

```
**Predicted  tensor**([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1], device='cuda:0')
    
    **Real Ground Truth**: tensor([0, 1, 2, 3, 4, 0, 1, 2, 3, 4, 0, 1, 2, 3, 4, 0, 1, 2, 3, 4, 0, 1, 2, 3,
        4, 0, 1, 2, 3, 4, 0, 1, 2, 3, 4, 0, 1, 2, 3, 4, 0, 1, 2, 3, 4, 0, 1, 2,
        3, 4], device='cuda:0')
```
 
 It only guesses label 1 some of the time, label 0 some of the time, and label 2, 3, or 4 some of the time. Average accuracy is always around 22%, so it can't learn.
 
```
 Predicted  tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0], device='cuda:0')
Real Ground Truth: tensor([0, 1, 2, 3, 4, 0, 1, 2, 3, 4, 0, 1, 2, 3, 4, 0, 1, 2, 3, 4, 0, 1, 2, 3,
        4, 0, 1, 2, 3, 4, 0, 1, 2, 3, 4, 0, 1, 2, 3, 4, 0, 1, 2, 3, 4, 0, 1, 2,
        3, 4], device='cuda:0')
```
        
Maybe my dataset is insufficient to learn from; maybe my deep learning model is not complex enough; maybe I am using Molgrid incorrectly. There's every possibility. I just want to make sure I didn't make a logical mistake with the molgrid part.

I am sharing the Google Colab notebook below. Am I making a logical mistake anywhere in terms of molgrid or anywhere?

https://drive.google.com/file/d/1q83UdU6RqKvCc4p_fK707EHHfiv3fX8C/view?usp=sharing

**Comment by dkoes** [comment]
You can see from the real labels that examples are being stratified.  I think you are in normal "trying to get my neural network to work" territory at this point.  
Make sure you can overfit to a single batch.
Reduce the learning rate.
Clip gradients.
Etc.

**Comment by drorhunvural** [comment]
Thank you for advices. I'll work on it. I guess stratify_receptor is grouping examples by their receptor identity and ensure an equal number of examples from each receptor in each minibatch. Do I need to keep stratify_receptor on True

**Comment by dkoes** [comment]
You should turn it off unless you will always have all five labels present for each receptor.
